zkvm/src/signature/prover.rs

```diff
@@ -0,0 +1,195 @@
+use curve25519_dalek::constants::RISTRETTO_BASEPOINT_POINT;
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use merlin::Transcript;
+use rand;
+use subtle::ConstantTimeEq;
+
+#[derive(Copy, Clone, Debug)]
+pub struct Nonce(Scalar);
+
+#[derive(Copy, Clone)]
+pub struct NoncePrecommitment([u8; 32]);
+
+#[derive(Copy, Clone, Debug)]
+pub struct NonceCommitment(RistrettoPoint);
+
+#[derive(Copy, Clone, Debug)]
+pub struct Siglet(Scalar);
+
+pub struct PartyAwaitingPrecommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    R_i: NonceCommitment,
+}
+
+pub struct PartyAwaitingCommitments {
+    transcript: Transcript,
+    multikey: Multikey,
+    x_i: PrivKey,
+    r_i: Nonce,
+    nonce_precommitments: Vec<NoncePrecommitment>,
+}
+
+pub struct PartyAwaitingSiglets {
+    multikey: Multikey,
+    c: Scalar,
+    nonce_commitments: Vec<NonceCommitment>,
+}
+
+impl NonceCommitment {
+    fn precommit(&self) -> NoncePrecommitment {
+        let mut h = Transcript::new(b"MuSig.nonce-precommit");
+        h.commit_point(b"R_i", &self.0.compress());
+        let mut precommitment = [0u8; 32];
+        h.challenge_bytes(b"precommitment", &mut precommitment);
+        NoncePrecommitment(precommitment)
+    }
+}
+
+impl PartyAwaitingPrecommitments {
+    pub fn new(
+        // The message `m` should already have been fed into the transcript
+        transcript: &Transcript,
+        x_i: PrivKey,
+        multikey: Multikey,
+    ) -> (Self, NoncePrecommitment) {
+        let mut rng_transcript = transcript.clone();
+        rng_transcript.commit_scalar(b"x_i", &x_i.0);
+        let mut rng = rng_transcript.build_rng().finalize(&mut rand::thread_rng());
+
+        // Generate ephemeral keypair (r_i, R_i). r_i is a random nonce.
+        let r_i = Nonce(Scalar::random(&mut rng));
+        // R_i = generator * r_i
+        let R_i = NonceCommitment(RISTRETTO_BASEPOINT_POINT * r_i.0);
+        // Make H(R_i)
+        let precommitment = R_i.precommit();
+
+        (
+            PartyAwaitingPrecommitments {
+                transcript: transcript.clone(),
+                multikey,
+                x_i,
+                r_i,
+                R_i,
+            },
+            precommitment,
+        )
+    }
+
+    pub fn receive_precommitments(
+        self,
+        nonce_precommitments: Vec<NoncePrecommitment>,
+    ) -> (PartyAwaitingCommitments, NonceCommitment) {
+        // Store received nonce precommitments in next state
+        (
+            PartyAwaitingCommitments {
+                transcript: self.transcript,
+                multikey: self.multikey,
+                x_i: self.x_i,
+                r_i: self.r_i,
+                nonce_precommitments,
+            },
+            self.R_i,
+        )
+    }
+}
+
+impl PartyAwaitingCommitments {
+    pub fn receive_commitments(
+        mut self,
+        nonce_commitments: Vec<NonceCommitment>,
```

Comment:
Not for this PR, but i wonder if there's a more clean way to pass around and collect every party's (pre)commitment? Maybe we need some thin Dealer implementation to orchestrate this?

Reply:
I considered having a dealer implementation, but each state change would just be a one-line map function (with the current implementation, such as in the tests), since the dealer doesn't actually have to hold on to any state... so I thought the logic was too simple to warrant making a dealer for. But if it makes the API easier to use, then it could be good!

Reply:
Maybe I'll get a better idea of what the API should look like in the next MuSig task, which is integrating this signing scheme into ZkVM!